lib/newspaper_works/data/work_derivatives.rb

```diff
@@ -0,0 +1,228 @@
+      class << self
+        attr_accessor :remap_names
+      end
+
+      # alternate constructor spelling:
+      def self.of(work, fileset = nil)
+        new(work, fileset)
+      end
+
+      # Adapt work and either specific or first fileset
+      def initialize(work, fileset = nil)
+        # adapted context usually work, may be string id of FileSet
+        @work = work
+        @fileset = fileset.nil? ? first_fileset : fileset
+        # computed name-to-path mapping, initially nil as sentinel for JIT load
+        @paths = nil
+        # assignments for attachment
+        @assigned = []
+        # un-assignments for deletion
+        @unassigned = []
+      end
+
+      # Assignment state
+      # @return [String] A label describing the state of assignment queues
+      def state
+        load_paths
+        return 'dirty' unless @unassigned.empty? && @assigned.empty?
+        return 'empty' if @paths.keys.empty?
+        'saved'
+      end
+
+      # Assign a path to assigned queue for attachment
+      # @param path [String] Path to source file
+      def assign(path)
+        path = normalize_path(path)
+        validate_path(path)
+        @assigned.push(path)
+      end
+
+      # Assign a destination name to unassigned queue for deletion -- OR --
+      #   remove a path from queue of assigned items
+      # @param name_or_id [String] Destination name (file extension),
+      #   or source path
+      def unassign(name)
+        # if name is queued path, remove from @assigned queue:
+        @assigned.delete(name) if @assigned.include?(name)
+        # if name is known destination name, remove
+        @unassigned.push(name) if exist?(name)
+      end
+
+      # commit pending changes to work files
+      #   beginning with removals, then with new assignments
+      def commit!
+        @unassigned.each { |name| delete(name) }
+        @assigned.each do |path|
+          attach(path, path_destination_name(path))
+        end
+        # reset queues after work is complete
+        @assigned = []
+        @unassigned = []
+      end
+
+      # attach a single derivative file to work
+      # @param file [String, IO] path to file or IO object
+      # @param name [String] destination name, usually file extension
+      def attach(file, name)
+        mkdir_pairtree
+        path = path_factory.derivative_path_for_reference(fileset, name)
+        # if file argument is path, copy file
+        if file.class == String
+          FileUtils.copy(file, path)
+        else
+          # otherwise, presume file is an IO, read, write it
+          #   note: does not close input file/IO, presume that is caller's
+          #   responsibility.
+          orig_pos = file.tell
+          file.seek(0)
+          File.open(path, 'w') { |dstfile| dstfile.write(file.read) }
+          file.seek(orig_pos)
+        end
+        # finally, reload @paths after mutation
+        load_paths
+      end
+
+      # attach a single derivative file to work
```

Comment:
Is this comment correct? Is this method attaching a single derivative? Possible mismatch with method name `delete`.

Reply:
Definitely a copy/paste issue.  Can correct comment.